### PR TITLE
NFT Offers Tx

### PR DESCRIPTION
**Pull request by benisgold**
Fixes APP-####

## What changed (plus any additional context for devs)
Enables accepting nft offers

### CODE REVIEW

**Comment by jxom** [comment]
> **New dependencies detected.** Learn more about [Socket for GitHub ↗︎](https://socket.dev?utm_medium=gh)
> 
> Packages	Version	New capabilities	Transitives	Size	Publisher
> [viem](https://socket.dev/npm/package/viem/overview/1.4.1)	[1.4.1](https://socket.dev/npm/package/viem/overview/1.4.1)	[network](https://socket.dev/npm/package/viem/issues/1.4.1?issue=networkAccess)	[`+8`](https://socket.dev/npm/package/viem/dependencies/1.4.1)	7.73 MB	[jmoxey](https://socket.dev/npm/user/jmoxey)

that’s my name 🥲

**Comment by benisgold** [comment]
> > **New dependencies detected.** Learn more about [Socket for GitHub ↗︎](https://socket.dev?utm_medium=gh)
> > Packages	Version	New capabilities	Transitives	Size	Publisher
> > [viem](https://socket.dev/npm/package/viem/overview/1.4.1)	[1.4.1](https://socket.dev/npm/package/viem/overview/1.4.1)	[network](https://socket.dev/npm/package/viem/issues/1.4.1?issue=networkAccess)	[`+8`](https://socket.dev/npm/package/viem/dependencies/1.4.1)	7.73 MB	[jmoxey](https://socket.dev/npm/user/jmoxey)
> 
> that’s my name 🥲

It's happening!!!

**Comment by skylarbarrera** [comment]
> > **New dependencies detected.** Learn more about [Socket for GitHub ↗︎](https://socket.dev?utm_medium=gh)
> 
> > 
> 
> > Packages	Version	New capabilities	Transitives	Size	Publisher
> 
> > [viem](https://socket.dev/npm/package/viem/overview/1.4.1)	[1.4.1](https://socket.dev/npm/package/viem/overview/1.4.1)	[network](https://socket.dev/npm/package/viem/issues/1.4.1?issue=networkAccess)	[`+8`](https://socket.dev/npm/package/viem/dependencies/1.4.1)	7.73 MB	[jmoxey](https://socket.dev/npm/user/jmoxey)
> 
> 
> 
> that’s my name 🥲

lfgggg

File: package.json [hunk 1]
```diff
@@ -212,7 +213,7 @@
     "react-native-aes-crypto": "rainbow-me/react-native-aes#65c49f7e70266615b2999eaa7db654d3fe4f2e3b",
     "react-native-ble-plx": "2.0.3",
     "react-native-bootsplash": "4.3.2",
-    "react-native-branch": "5.3.1",
+    "react-native-branch": "5.9.0",
```
Comment (skylarbarrera) [inline package.json:216]: gotta revert this bump, not working on android

File: src/App.js [hunk 2]
```diff
@@ -277,6 +287,25 @@ function App() {
...
+        apiKey: RESERVOIR_API_KEY,
+      },
+      {
+        id: 137,
+        baseApiUrl: 'https://api-polygon.reservoir.tools',
+        active: true,
+        apiKey: RESERVOIR_API_KEY,
+      },
+    ],
+    logLevel: 4,
```
Comment (skylarbarrera) [inline src/App.js:306]: up to you here but curious if we want this in prod vs dev etc

File: ios/Rainbow.xcodeproj/project.pbxproj [hunk 3]
```diff
@@ -9,6 +9,7 @@
 /* Begin PBXBuildFile section */
 		0299CE7B2886202800B5C7E7 /* NotificationService.m in Sources */ = {isa = PBXBuildFile; fileRef = 0299CE7A2886202800B5C7E7 /* NotificationService.m */; };
 		0299CE7F2886202800B5C7E7 /* ImageNotification.appex in Embed App Extensions */ = {isa = PBXBuildFile; fileRef = 0299CE772886202800B5C7E7 /* ImageNotification.appex */; settings = {ATTRIBUTES = (RemoveHeadersOnCopy, ); }; };
+		0A75490AB3C4D644E34D9BE0 /* libPods-PriceWidgetExtension.a in Frameworks */ = {isa = PBXBuildFile; fileRef = 2D9D6426C5AE0E210FF6C21A /* libPods-PriceWidgetExtension.a */; };
```
Comment (skylarbarrera) [inline ios/Rainbow.xcodeproj/project.pbxproj:12]: I don't ~think we need any changes to this file

File: src/components/coin-row/FastTransactionStatusBadge.tsx [hunk 4]
```diff
@@ -121,6 +121,15 @@ const StatusProps = {
       marginTop: ios ? -3 : -5,
     },
   },
+  [TransactionStatusTypes.sold]: {
+    name: 'sunflower',
```
Comment (skylarbarrera) [inline src/components/coin-row/FastTransactionStatusBadge.tsx:125]: 🌻  ??

File: src/handlers/nftOffers.ts [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+  approval: TxData | undefined,
+  sale: TxData | undefined
+): Promise<string | null> => {
+  const provider = await getProviderForNetwork(offer.network);
+  if (!sale) {
+    if (offer.marketplace.name !== 'Blur') {
+      // expecting sale tx data for all marketplaces except Blur
+      logger.warn('No sale tx data for NFT Offer');
+    }
+    return '1000000';
```
Comment (skylarbarrera) [inline src/handlers/nftOffers.ts:77]: can we add these values to eth_units

File: src/handlers/nftOffers.ts [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+  }
+  if (!approval) {
+    return await estimateGasWithPadding(sale, null, null, provider);
+  }
+  if (offer.network !== Network.mainnet) {
+    // rough gas estimate
+    return '1500000';
```
Comment (skylarbarrera) [inline src/handlers/nftOffers.ts:84]: eth_units

File: src/screens/NFTSingleOfferSheet/index.tsx [hunk 7]
```diff
@@ -136,16 +210,228 @@ export function NFTSingleOfferSheet() {
...
-  const feesPercentage = Math.floor(offer.feesPercentage * 10) / 10;
-  const royaltiesPercentage = Math.floor(offer.royaltiesPercentage * 10) / 10;
+  // estimate gas
+  useEffect(() => {
+    if (!isReadOnlyWallet && !isExpired) {
+      (async () => {
+        const privateKey = await loadPrivateKey(accountAddress, false);
+        // @ts-ignore
+        const account = privateKeyToAccount(privateKey);
+        const reservoirSigner = createWalletClient({
```
Comment (skylarbarrera) [inline src/screens/NFTSingleOfferSheet/index.tsx:221]: I would love this in a function vs straight in the useEffect

File: src/screens/NFTSingleOfferSheet/index.tsx [hunk 8]
```diff
@@ -136,16 +210,228 @@ export function NFTSingleOfferSheet() {
...
-    : undefined;
-  const buttonColorFallback = useForegroundColor('accent');
 
-  const feesPercentage = Math.floor(offer.feesPercentage * 10) / 10;
-  const royaltiesPercentage = Math.floor(offer.royaltiesPercentage * 10) / 10;
+  // estimate gas
+  useEffect(() => {
+    if (!isReadOnlyWallet && !isExpired) {
+      (async () => {
+        const privateKey = await loadPrivateKey(accountAddress, false);
```
Comment (skylarbarrera) [inline src/screens/NFTSingleOfferSheet/index.tsx:218]: this is going to trigger faceId on mount, we should be able to pass in a read only signer for gas estimations iirc

File: src/screens/NFTSingleOfferSheet/index.tsx [hunk 9]
```diff
@@ -136,16 +210,228 @@ export function NFTSingleOfferSheet() {
...
+        // @ts-ignore
+        const account = privateKeyToAccount(privateKey);
+        const reservoirSigner = createWalletClient({
+          account,
+          chain: mainnet,
+          transport: http(getNetworkObj(offer.network as Network).rpc),
+        });
+        setSigner(reservoirSigner);
```
Comment (skylarbarrera) [inline src/screens/NFTSingleOfferSheet/index.tsx:226]: would prefer to do read-only and then not save signer in state, don't wanna be keeping objects with keys around if we don't have to

File: src/screens/NFTSingleOfferSheet/index.tsx [hunk 10]
```diff
@@ -136,16 +210,228 @@ export function NFTSingleOfferSheet() {
...
+  // estimate gas
+  useEffect(() => {
+    if (!isReadOnlyWallet && !isExpired) {
+      (async () => {
+        const privateKey = await loadPrivateKey(accountAddress, false);
+        // @ts-ignore
+        const account = privateKeyToAccount(privateKey);
+        const reservoirSigner = createWalletClient({
+          account,
+          chain: mainnet,
```
Comment (skylarbarrera) [inline src/screens/NFTSingleOfferSheet/index.tsx:223]: only mainnet?

**Review by skylarbarrera — CHANGES_REQUESTED** [review]
smol clean up

File: package.json (hunk 1 above)
Comment (benisgold) [reply package.json:216]: oops not sure how that got in

File: src/components/coin-row/FastTransactionStatusBadge.tsx (hunk 4 above)
Comment (benisgold) [reply src/components/coin-row/FastTransactionStatusBadge.tsx:125]: yes

Comment (benisgold) [reply src/components/coin-row/FastTransactionStatusBadge.tsx:125]: fight me

File: ios/Rainbow.xcodeproj/project.pbxproj (hunk 3 above)
Comment (benisgold) [reply ios/Rainbow.xcodeproj/project.pbxproj:12]: i didnt do anything it was autogenerated afaik

File: src/App.js [hunk 11]
```diff
@@ -358,6 +359,7 @@ function Root() {
         // for failure, continue to rest of the app for now
         setInitializing(false);
       });
+    initializeReservoirClient();
```
Comment (skylarbarrera) [inline src/App.js:362]: still unsure this needs to be here specifically but cool 4 now

File: src/screens/NFTSingleOfferSheet/index.tsx [hunk 12]
```diff
@@ -136,16 +209,234 @@ export function NFTSingleOfferSheet() {
...
+    network,
+    offer,
+    startPollingGasFees,
+    updateTxFee,
+  ]);
+
+  // estimate gas
+  useEffect(() => {
+    if (!isReadOnlyWallet && !isExpired) {
+      estimateGas();
```
Comment (skylarbarrera) [inline src/screens/NFTSingleOfferSheet/index.tsx:276]: clean asf

**Review by skylarbarrera — APPROVED** [review]
lets ride

**Pull request merged**